src/service/server.rs

```diff
@@ -19,60 +20,68 @@ use crate::util::error::{quit_error, ErrorHints};
-
+pub async fn service(bind_addr: SocketAddr, configs: Vec<Arc<Config>>) -> Result<(), ()> {
     // Listen for new connections
-    let listener = TcpListener::bind(config.public.address)
-        .await
-        .map_err(|err| {
-            quit_error(
-                anyhow!(err).context("Failed to start proxy server"),
-                ErrorHints::default(),
-            );
-        })?;
+    let listener = TcpListener::bind(bind_addr).await.map_err(|err| {
+        quit_error(
+            anyhow!(err).context("Failed to start proxy server"),
+            ErrorHints::default(),
+        );
+    })?;
 
-    info!(
-        target: "lazymc",
-        "Proxying public {} to server {}",
-        config.public.address, config.server.address,
-    );
+    let mut servers = HashMap::new();
 
-    if config.lockout.enabled {
-        warn!(
+    for config in configs.iter() {
+        // Load server state
+        let server = Arc::new(Server::default());
+        servers.insert(
+            config.public.address.clone(),
+            (config.clone(), server.clone()),
+        );
+
+        info!(
             target: "lazymc",
-            "Lockout mode is enabled, nobody will be able to connect through the proxy",
+            "Proxying public {} to server {}",
+            config.public.address, config.server.address,
         );
-    }
 
-    // Spawn services: monitor, signal handler
-    tokio::spawn(service::monitor::service(config.clone(), server.clone()));
-    tokio::spawn(service::signal::service(config.clone(), server.clone()));
+        if config.lockout.enabled {
+            warn!(
+                target: "lazymc",
+                "Lockout mode is enabled, nobody will be able to connect through the proxy",
+            );
+        }
+
+        // Spawn services: monitor, signal handler
+        tokio::spawn(service::monitor::service(config.clone(), server.clone()));
+        tokio::spawn(service::signal::service(config.clone(), server.clone()));
+
+        // Initiate server start
+        if config.server.wake_on_start {
+            Server::start(config.clone(), server.clone(), None).await;
+        }
 
-    // Initiate server start
-    if config.server.wake_on_start {
-        Server::start(config.clone(), server.clone(), None).await;
+        // Spawn additional services: probe and ban manager
+        tokio::spawn(service::probe::service(config.clone(), server.clone()));
+        tokio::task::spawn_blocking({
+            let (config, server) = (config.clone(), server.clone());
+            || service::file_watcher::service(config, server)
+        });
     }
 
-    // Spawn additional services: probe and ban manager
-    tokio::spawn(service::probe::service(config.clone(), server.clone()));
-    tokio::task::spawn_blocking({
-        let (config, server) = (config.clone(), server.clone());
-        || service::file_watcher::service(config, server)
-    });
+    let servers_arc = Arc::new(servers);
```

Comment:
Is `Arc<HashMap<(Arc<Config>, Arc<Server>)>>` a logical data structure here, are there not too many `Arc`s?

Reply:
Yes I think that is too complex.

See my other comment in: `src/action/start.rs`

Reply:
Resolved with a new type, see my comment https://github.com/timvisee/lazymc/pull/42#discussion_r1103622867